### PR TITLE
add warning message when delete to release-3.5

### PR DESCRIPTION
**Pull request by kkkkun**
Fixes https://github.com/etcd-io/etcd/issues/13705

Base PR: https://github.com/etcd-io/etcd/pull/13729 
Change commit branch from main to release-3.5

### CODE REVIEW

**Review by ptabor — CHANGES_REQUESTED** [review]
If I understand correctly, the only case we want to issue a warning is case
that happens in line 72 of the original code: 

```
opts = append(opts, clientv3.WithRange(args[1]))
```

i.e. the flags like `--prefix', `--from-key` or (to be added) `--range` are not given... 

The code should IMHO print warning only iff: 
  - none of the flag is given
  - the command has 2 arguments. 
  
  BTW: I would still recommend submiting a PR to main adding the `--range` flag and warning .... and backport it to 3.5 later.

**Comment by kkkkun** [comment]
> If I understand correctly, the only case we want to issue a warning is case that happens in line 72 of the original code:
> 
> ```
> opts = append(opts, clientv3.WithRange(args[1]))
> ```
> 
> i.e. the flags like `--prefix', `--from-key`or (to be added)`--range` are not given...
> 
> The code should IMHO print warning only iff:
> 
> * none of the flag is given
> * the command has 2 arguments.
> 
> BTW: I would still recommend submiting a PR to main adding the `--range` flag and warning .... and backport it to 3.5 later.

According to your advicer. I sumbit this to release-3.5
and submit  https://github.com/etcd-io/etcd/pull/13747  to main for 3.6 or later ?

File: etcdctl/ctlv3/command/del_command.go [hunk 1]
```diff
@@ -70,6 +72,9 @@ func getDelOp(args []string) (string, []clientv3.OpOption) {
 			cobrautl.ExitWithError(cobrautl.ExitBadArgs, fmt.Errorf("too many arguments, only accept one argument when `--prefix` or `--from-key` is set"))
 		}
 		opts = append(opts, clientv3.WithRange(args[1]))
+		ignoreWarn = false
+	} else if delPrefix || delFromKey {
+		ignoreWarn = false
```
Comment (ptabor) [inline etcdctl/ctlv3/command/del_command.go:77]: It's not needed.... We don't need IMHO warning if user explicitly stated '--from-key' or `--prefix`.
This logic would print the warning in such situation as well.

File: etcdctl/ctlv3/command/del_command.go [hunk 2]
```diff
@@ -91,5 +96,9 @@ func getDelOp(args []string) (string, []clientv3.OpOption) {
 		opts = append(opts, clientv3.WithFromKey())
 	}
 
+	if !ignoreWarn {
+		fmt.Fprintf(os.Stderr, "Warning: will delete servarl keys.")
```
Comment (ptabor) [inline etcdctl/ctlv3/command/del_command.go:100]: I would state it as:
```
Warning: This operations is implicitly deleting range of keys. Please add flag `--range` to make it explicit. Will sleep in v3.5. 
```

**Review by ptabor — CHANGES_REQUESTED** [review]
I think it does not work....
It prints warning only

File: etcdctl/ctlv3/command/del_command.go (hunk 2 above)
Comment (kkkkun) [reply etcdctl/ctlv3/command/del_command.go:100]: > I would state it as:
> 
> ```
> Warning: This operations is implicitly deleting range of keys. Please add flag `--range` to make it explicit. Will sleep in v3.5. 
> ```


This PR commit to release-3.5?

Should we chang to this
```
Warning: This operations is implicitly deleting range of keys. 
```
When commit to main:
```
> Warning: This operations is implicitly deleting range of keys. Please add flag `--range` to make it explicit. Will sleep in v3.6. 
```

File: etcdctl/ctlv3/command/del_command.go [hunk 3]
```diff
@@ -70,6 +72,9 @@ func getDelOp(args []string) (string, []clientv3.OpOption) {
 			cobrautl.ExitWithError(cobrautl.ExitBadArgs, fmt.Errorf("too many arguments, only accept one argument when `--prefix` or `--from-key` is set"))
 		}
 		opts = append(opts, clientv3.WithRange(args[1]))
+		ignoreWarn = false
```
[the remark this answers is not in the document]
Comment (kkkkun) [reply etcdctl/ctlv3/command/del_command.go:75]: OK . I do not understand before

**Comment by ahrtr** [comment]
Why the warning message  in this PR is different from the message in [13747](https://github.com/etcd-io/etcd/pull/13747) ?

This is a kind of enhancement, so in my opinion it should NOT be backported to 3.5.  However it's just a very minor change, I will not insist on this if other maintainers/reviewers approve this PR.

**Comment by kkkkun** [comment]
> Why the warning message in this PR is different from the message in [13747](https://github.com/etcd-io/etcd/pull/13747) ?
> 
> This is a kind of enhancement, so in my opinion it should NOT be backported to 3.5. However it's just a very minor change, I will not insist on this if other maintainers/reviewers approve this PR.

Copy from https://github.com/etcd-io/etcd/issues/13705#issuecomment-1047168535
It‘s just a warning message.
```
 In 3.5 we could add a warning on stderr.
 In 3.6 we could add support for --range and sleep (2s) if the --range is not given:
 a) let the operation be cancelled by human operator
 b) trigger attention of the script owners to start using the new flag.
```

File: etcdctl/ctlv3/command/del_command.go [hunk 4]
```diff
@@ -39,6 +42,7 @@ func NewDelCommand() *cobra.Command {
 	cmd.Flags().BoolVar(&delPrefix, "prefix", false, "delete keys with matching prefix")
 	cmd.Flags().BoolVar(&delPrevKV, "prev-kv", false, "return deleted key-value pairs")
 	cmd.Flags().BoolVar(&delFromKey, "from-key", false, "delete keys that are greater than or equal to the given key using byte compare")
+	cmd.Flags().BoolVar(&delRange, "range", false, "delete range of keys")
```
Comment (ahrtr) [inline etcdctl/ctlv3/command/del_command.go:45]: If I understand @ptabor 's comment correctly, we don't need add any flag for 3.5. Instead, we just need to add a warning message.

Comment (kkkkun) [reply etcdctl/ctlv3/command/del_command.go:45]: https://github.com/etcd-io/etcd/pull/13729#discussion_r814548212
In this comment. ptabor support add a flag `--range` to ingore this message.

File: etcdctl/ctlv3/command/del_command.go [hunk 5]
```diff
@@ -70,6 +74,10 @@ func getDelOp(args []string) (string, []clientv3.OpOption) {
 			cobrautl.ExitWithError(cobrautl.ExitBadArgs, fmt.Errorf("too many arguments, only accept one argument when `--prefix` or `--from-key` is set"))
 		}
 		opts = append(opts, clientv3.WithRange(args[1]))
+		if !delRange {
+			fmt.Fprintln(os.Stderr, "Warning: This operations is implicitly deleting range of keys. Please add flag `--range` to make it explicit. Will sleep in v3.5.")
```
Comment (ahrtr) [inline etcdctl/ctlv3/command/del_command.go:78]: Please use the same warning message as the main branch.

Comment (kkkkun) [reply etcdctl/ctlv3/command/del_command.go]: Fixed!

**Review by ahrtr — APPROVED** [review]
LGTM

cc @serathius @spzala @ptabor  Please see my comment [issuecomment-1126916984](https://github.com/etcd-io/etcd/pull/13748#issuecomment-1126916984).

**Review by ptabor — CHANGES_REQUESTED** [review]
This PR adds Sleep(2sec) to 3.5, so has potential to break customer's workflows thus cannot be accepted.

I'm for submitting just warning+flag to 3.5, to let customer's prepare for 3.6, but the message needs to be adjusted.

**Comment by kkkkun** [comment]
> This PR adds Sleep(2sec) to 3.5, so has potential to break customer's workflows thus cannot be accepted.
> 
> I'm for submitting just warning+flag to 3.5, to let customer's prepare for 3.6, but the message needs to be adjusted.

The message has been adjusted. And Sleep(2sec) has been removed.

File: etcdctl/ctlv3/command/del_command.go [hunk 6]
```diff
@@ -16,16 +16,18 @@ package command
 
 import (
 	"fmt"
-
 	"github.com/spf13/cobra"
-	"go.etcd.io/etcd/client/v3"
+	clientv3 "go.etcd.io/etcd/client/v3"
 	"go.etcd.io/etcd/pkg/v3/cobrautl"
+	"os"
+	"time"
```
Comment (ahrtr) [inline etcdctl/ctlv3/command/del_command.go:23]: Please try to group the import entries properly. The build-in package `os`, `time` and `fmt` should be grouped together and above other packages.

**Comment by ahrtr** [comment]
> This PR adds Sleep(2sec) to 3.5, so has potential to break customer's workflows thus cannot be accepted.
> 
> I'm for submitting just warning+flag to 3.5, to let customer's prepare for 3.6, but the message needs to be adjusted.

@ptabor was requesting to remove the `sleep`, and I agree with that.  Please also remove "`Will sleep in v3.5.`" from the message as well, or change it to "Will sleep for a while in v3.6"?.  cc @spzala for the professional opinion for the English syntax.

**Comment by ahrtr** [comment]
Please sign off the commit. 

```
git commit --signoff --amend
```

**Comment by ahrtr** [comment]
Looks good to me now, @ptabor  Do you have any comments?

**Comment by ahrtr** [comment]
Just to be clearer, this PR is a partially backporting of https://github.com/etcd-io/etcd/issues/13705

This is a minor enhancement for the `etcdctl del ` command. Usually we only backport bug fix, so we shouldn't backport this PR. But since it's just a minor enhancement and it's useful, so I will not object if other maintainers agree to backport it. 

What's your opinion? @ptabor @spzala @serathius  If there is no response, then I assume there is no interest on this PR for 3.5. Then we can close it.

File: etcdctl/ctlv3/command/del_command.go [hunk 7]
```diff
@@ -70,6 +73,9 @@ func getDelOp(args []string) (string, []clientv3.OpOption) {
 			cobrautl.ExitWithError(cobrautl.ExitBadArgs, fmt.Errorf("too many arguments, only accept one argument when `--prefix` or `--from-key` is set"))
 		}
 		opts = append(opts, clientv3.WithRange(args[1]))
+		if !delRange {
+			fmt.Fprintln(os.Stderr, "Warning: This operations is implicitly deleting range of keys. Please add flag `--range` to make it explicit. Will sleep for a while in v3.6.")
+		}
```
Comment (spzala) [inline etcdctl/ctlv3/command/del_command.go:78]: I agree with @ahrtr that `Will sleep for a while...` is difficult to interpret. Sorry, I took longer to review, but I would suggest something like these: 
`In etcd v3.6, along with this warning, the delete operation will be suspended for a few seconds to provide the user time to verify range.`    
 or 
`In etcd v3.6, the operation will be suspended for a few seconds to provide the user time to verify range.`
 or 
`In v3.6, the operation will be suspended for a few seconds to provide the user time to verify range.`
Thanks @kkkkun

**Comment by spzala** [comment]
> Just to be clearer, this PR is a partially backporting of #13747
> 
> This is a minor enhancement for the `etcdctl del ` command. Usually we only backport bug fix, so we shouldn't backport this PR. But since it's just a minor enhancement and it's useful, so I will not object if other maintainers agree to backport it.
> 
> What's your opinion? @ptabor @spzala @serathius If there is no response, then I assume there is no interest on this PR for 3.5. Then we can close it.

@ahrtr I agree with your thoughts here. I am okay with backporting. Thanks!

File: etcdctl/ctlv3/command/del_command.go (hunk 7 above)
Comment (kkkkun) [reply etcdctl/ctlv3/command/del_command.go:78]: ok, I  would change to the second
`In etcd v3.6, the operation will be suspended for a few seconds to provide the user time to verify range.`

Comment (kkkkun) [reply etcdctl/ctlv3/command/del_command.go:78]: Done！

**Comment by ahrtr** [comment]
Thanks @kkkkun

**Pull request merged**